### PR TITLE
Add support for including optional attendees in meeting times

### PR DESCRIPTION
**Pull request by ALai2**
Added new basic functionality to query function so that if one or more time slots exist in which both optional and mandatory attendees can attend those time slots will be returned. Otherwise, return the time slots that fit only the mandatory attendees. I also had to add a check for null in getOptionalAttendees because using the already written code that was given gives an error for null pointer exception when running the web application.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java [hunk 1]
```diff
@@ -46,6 +46,9 @@ public MeetingRequest(Collection<String> attendees, long duration) {
    * Returns a read-only copy of the people who are optional to attend this meeting.
    */
   public Collection<String> getOptionalAttendees() {
+    if (optional_attendees == null) {
+      return Collections.unmodifiableCollection(new HashSet<>());
```
Comment (linxiz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java]: You could directly use Collections.emptySet();

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -47,7 +47,7 @@
     Collections.sort(outOfRange, TimeRange.ORDER_BY_START);
 
     // iterate through and add timeranges to collection of available times
-    Collection<TimeRange> availableTimes = new ArrayList<>();
+    ArrayList<TimeRange> availableTimes = new ArrayList<>();
```
Comment (linxiz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: nit: Is this change necessary?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -66,6 +66,42 @@
       availableTimes.add(TimeRange.fromStartEnd(startPointer, TimeRange.END_OF_DAY, true));
     }
 
-    return availableTimes;
+    Collection<TimeRange> optionalAvailableTimes = new ArrayList<>();
+    if (request.getOptionalAttendees().size() > 0) {
+      // get available times for optional attendees
+      MeetingRequest requestOptionalTimes = new MeetingRequest(request.getOptionalAttendees(), request.getDuration());
```
Comment (linxiz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:72]: nit: you might want to rename it to optionalTimesRequest

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -66,6 +66,42 @@
...
+      Collection<TimeRange> optionalTimes = query(events, requestOptionalTimes);
+
+      // find timeranges where available times for mandatory and optional attendees overlap
+      int j = 0;
+      for (TimeRange optionalTime: optionalTimes) {
+        for (int i = j; i < availableTimes.size(); i++) {
+          j = i;
```
Comment (linxiz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:79]: If I read it correctly, the size of availableTimes is never changed during these loops, thus j is always size() - 1 after the loop for the 1st optional time. Do you intend to keep comparing only to the last available time for the optional times starting from the 2nd optional time on?

**Review by linxiz — COMMENTED** [review]
Could you add corresponding test?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -66,6 +66,42 @@
       availableTimes.add(TimeRange.fromStartEnd(startPointer, TimeRange.END_OF_DAY, true));
```
Comment (linxiz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:66]: I just noticed that END_OF_DAY is actually 23:59, do you actually want to use WHOLE_DAY for this calculation instead? e.g. is a whole_day event supported?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (ALai2) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: Yes, when I use Collection<TimeRange> availableTimes it gives me an error for .get() when I try to do TimeRange currentAvailableTime = availableTimes.get(i); at line 80.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (ALai2) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:79]: The variable j is there to prevent available mandatory times from being seen repeatedly. I break out of the inner for loop when I want to go to the next optional time and so I want to start at the available mandatory time that I left off in the previous loop. If j is size()-1 after the loop for the 1st optional time that means the all the available mandatory times are either before or contained within that 1st optional time. Then it's right to want to compare only the last available mandatory time for the rest of the optional times since the times are sorted.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (ALai2) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:66]: Using WHOLE_DAY there would be applicable only if there were no events that mandatory attendees attend. Lines 64-67 check to see if after the last mandatory attended event there is a gap big enough for an event between startPointer and the end of the day. By the algorithm if outOfRange has at least one event, than startPointer by lines 64-67 would be eventTime.end(), or the end of the last event, then I would not want to use WHOLE_DAY. If outOfRange has no events, then startPointer would be START_OF_DAY, then yes, I would want the calculation to give back WHOLE_DAY.

**Comment by ALai2** [comment]
> Could you add corresponding test?

The corresponding tests for optional attendee functionality were added in a previous pull request (PR#23). Unless you mean for me to add more tests?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (linxiz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:66]: Could you add tests for whole day events or events that end after 23:59?

Comment (ALai2) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:66]: I added three tests: wholeDayEventBothAttend(), PersonAHasWholeDayEvent() and EventEndsAfterEndOfDay(). The first two tests for whole day events and the third is a test for an event that ends after 23:59.

Comment (linxiz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:66]: Could you also add test for querying a DURATION_30_MINUTES event when there is an existing event end at TIME_1130PM? Will there be available option?
|--------A--------|         event ends at 23:30
|---------------------|   a day

**Review by linxiz — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (ALai2) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:66]: I added the test, because it includes the end of day there will be an available option.
